### PR TITLE
Updating integration tests to test both API versions - v1beta1 and 3

### PR DESCRIPTION
**Pull request by nikhiljindal**
https://github.com/GoogleCloudPlatform/kubernetes/issues/5475

### CODE REVIEW

**Comment by bgrant0607** [comment]
```
--- FAIL: TestExampleObjectSchemas (0.23s)
	examples_test.go:159: ../cmd/integration/v1beta1-controller.json does not have a test case defined
	examples_test.go:159: ../cmd/integration/v1beta3-controller.json does not have a test case defined
	examples_test.go:175: Expected 1 examples, Got 0
FAIL
coverage: 0.0% of statements
FAIL	github.com/GoogleCloudPlatform/kubernetes/examples	0.380s
```

**Comment by nikhiljindal** [comment]
Thats a cool test that I didnt know existed!
Fixed.

File: cmd/integration/integration.go [hunk 1]
```diff
@@ -373,55 +376,55 @@ func runReplicationControllerTest(c *client.Client) {
...
-	// TODO: this is not namespace aware
-	err = c.Get().RequestURI(svc.SelfLink).Do().Into(&svc)
+	err = c.Get().RequestURI(svc.SelfLink).Do().Into(svc)
 	if err != nil {
 		glog.Fatalf("Failed listing service with supplied self link '%v': %v", svc.SelfLink, err)
 	}
 
-	var svcList api.ServiceList
-	err = c.Get().NamespaceIfScoped(namespace, len(namespace) > 0).Resource("services").Do().Into(&svcList)
+	svcList, err := services.List(labels.Everything())
```
Comment (bgrant0607) [inline cmd/integration/integration.go]: Why did you remove the namespacing?

File: cmd/integration/integration.go [hunk 2]
```diff
@@ -448,29 +451,28 @@ func runSelfLinkTestOnNamespace(c *client.Client, namespace string) {
...
-			TypeMeta: api.TypeMeta{
-				APIVersion: latest.Version,
-			},
-			ObjectMeta: api.ObjectMeta{
-				Name: "atomicservice",
-				Labels: map[string]string{
-					"name": "atomicService",
-				},
+	svcBody := api.Service{
+		TypeMeta: api.TypeMeta{
```
Comment (bgrant0607) [inline cmd/integration/integration.go]: Why do you set the API version in TypeMeta for some objects, but not for others?

File: cmd/integration/integration.go (hunk 1 above)
Comment (bgrant0607) [inline cmd/integration/integration.go]: Nevermind.

Comment (derekwaynecarr) [inline cmd/integration/integration.go]: It looks like its still there, but services is now ServiceNamespacer that has already been scoped to the namespace in question.

File: cmd/kubernetes/kubernetes.go [hunk 3]
```diff
@@ -86,7 +86,6 @@ func runApiServer(cl *client.Client, etcdClient tools.EtcdClient, addr net.IP, p
 
 	// Create a master and install handlers into mux.
 	m := master.New(&master.Config{
-		Client:     cl,
```
Comment (bgrant0607) [inline cmd/kubernetes/kubernetes.go]: Why this change?

File: cmd/kube-apiserver/app/server.go [hunk 4]
```diff
@@ -229,7 +229,6 @@ func (s *APIServer) Run(_ []string) error {
 	admissionController := admission.NewFromPlugins(client, admissionControlPluginNames, s.AdmissionControlConfigFile)
 
 	config := &master.Config{
-		Client:                 client,
```
Comment (bgrant0607) [inline cmd/kube-apiserver/app/server.go]: Why this change?

File: cmd/integration/integration.go (hunk 2 above)
Comment (nikhiljindal) [inline cmd/integration/integration.go]: I havent changed the objects at all in this PR. I am just changing the way they are used.
I can add API Version to all objects, if you want.

File: cmd/kubernetes/kubernetes.go (hunk 3 above)
Comment (nikhiljindal) [inline cmd/kubernetes/kubernetes.go]: Master is not using it all.

File: cmd/kube-apiserver/app/server.go (hunk 4 above)
Comment (nikhiljindal) [inline cmd/kube-apiserver/app/server.go]: Same

**Comment by nikhiljindal** [comment]
Replied to all comments. PTAL.

File: pkg/master/master.go [hunk 5]
```diff
@@ -65,7 +65,6 @@ import (
 
 // Config is a structure used to configure a Master.
 type Config struct {
-	Client            *client.Client
```
Comment (bgrant0607) [inline pkg/master/master.go]: Again, why this change?

File: cmd/kubernetes/kubernetes.go (hunk 3 above)
Comment (bgrant0607) [inline cmd/kubernetes/kubernetes.go]: Why is it still a parameter to the function, then?

File: cmd/integration/integration.go (hunk 2 above)
Comment (bgrant0607) [inline cmd/integration/integration.go]: Ok.

File: pkg/master/master.go (hunk 5 above)
Comment (bgrant0607) [inline pkg/master/master.go]: Nevermind. I see your answer above now.

File: test/integration/client_test.go [hunk 6]
```diff
@@ -50,7 +50,6 @@ func TestClient(t *testing.T) {
 	defer s.Close()
 
 	m = master.New(&master.Config{
```
Comment (bgrant0607) [inline test/integration/client_test.go]: See the client_test.go change in #5085 to enable v1beta3 testing.
cc @smarterclayton

File: cmd/kubernetes/kubernetes.go (hunk 3 above)
Comment (nikhiljindal) [inline cmd/kubernetes/kubernetes.go]: Thanks for pointing that out. Removed.

**Comment by bgrant0607** [comment]
LGTM. Waiting for tests.

**Comment by nikhiljindal** [comment]
Shippable passed :)

File: pkg/client/services.go [hunk 7]
```diff
@@ -77,7 +77,13 @@ func (c *services) Get(name string) (result *api.Service, err error) {
 // Create creates a new service.
 func (c *services) Create(svc *api.Service) (result *api.Service, err error) {
 	result = &api.Service{}
-	err = c.r.Post().Namespace(c.ns).Resource("services").Body(svc).Do().Into(result)
+	// v1beta3 does not allow POST without a namespace.
+	needNamespace := !api.PreV1Beta3(c.r.APIVersion())
```
Comment (smarterclayton) [inline pkg/client/services.go]: Why is this needed?  I think this is a bug in higher level code, and doesn't look right here.

Comment (nikhiljindal) [inline pkg/client/services.go]: If not here, then we will need such conditions everywhere at higher level code.
```
If (apiVersion = v1beta3 && ns == "") {
  ns = api.NamespaceDefault
}
servicesInterface = client.services(ns)
```
Is that better?

Comment (smarterclayton) [inline pkg/client/services.go]: I'm saying that anyone who has an empty string either got it from client input (in which case the server will catch it) or it will be caught lower in the code base.  No one should be passing an empty namespace, ever.  I don't think we need to test in the client (I have a pull incoming which normalizes all our clients now that the server is normalized).

----- Original Message -----

Comment (nikhiljindal) [inline pkg/client/services.go]: ok. What do you mean by "normalizing the server and the client"?

Comment (smarterclayton) [inline pkg/client/services.go]: The RESTStorage objects and the api_installer now properly check a lot of the conditions. If there is client code that is allowing Namespace to be wrong when it passes to the Services code, we should either fix the client code or reject it on the server (the standard api/rest.go code should do that now).

Opened https://github.com/GoogleCloudPlatform/kubernetes/pull/5738 to clean some of this up.

----- Original Message -----

Comment (nikhiljindal) [inline pkg/client/services.go]: ok. Thanks for the explanation!
I looked at the PR and it looks good.
Thanks for working on it.
